CMakeLists.txt

```
@@ -26,82 +24,165 @@ if(HIOP_USE_MPI)
+                            # float to double
+)
+
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+    message("Using extra GNU-specific compiler warnings...")
+    target_compile_options(hiop_warnings
+        INTERFACE
+            -Wmisleading-indentation    # Warn on indented blocks
+                                        # that are not really blocks
+            -Wduplicated-cond           # Warn on if/else chain with
+                                        # duplicated conditions
+            -Wduplicated-branches       # Warn on if/else chains with
+                                        # duplicated code
+            -Wuseless-cast              # Warn when casting to the same type
+    )
+endif()
+
+if(HIOP_DEVELOPER_MODE)
+    target_link_libraries(hiop_math INTERFACE hiop_options hiop_warnings)
+endif()
+
 if(HIOP_USE_GPU)
+  include(CheckLanguage)
+  enable_language(CUDA)
+  check_language(CUDA)
+  if(NOT DEFINED CMAKE_CUDA_STANDARD)
+    set(CMAKE_CUDA_STANDARD 11)
+    set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+  endif()
+
+  if(HIOP_NVCC_ARCH)
+      set(CMAKE_CUDA_FLAGS "-arch=${HIOP_NVCC_ARCH}")
+      message(STATUS "Using CUDA arch ${HIOP_NVCC_ARCH}")
+  else()
+      set(CMAKE_CUDA_FLAGS "-arch=sm_35")
+      message(STATUS "Using CUDA arch sm_35")
+  endif()
+  set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} --expt-extended-lambda")
+
   # for now we rely on MAGMA for GPUs computations
-  set(MAGMA_LIBRARIES "-lmagma")
+  target_link_libraries(hiop_magma INTERFACE "-lmagma")
   if(HIOP_MAGMA_DIR)
-    set(MAGMA_LIBRARIES "-L${HIOP_MAGMA_DIR}/lib ${MAGMA_LIBRARIES}")
-    set(MAGMA_INCLUDE "${HIOP_MAGMA_DIR}/include")
+    target_link_libraries(hiop_magma INTERFACE "-L${HIOP_MAGMA_DIR}/lib")
+    target_include_directories(hiop_magma INTERFACE "${HIOP_MAGMA_DIR}/include")
   endif(HIOP_MAGMA_DIR)
 
-  set(HIOP_CUDA_LIBRARIES "-lculibos -lcublas -lcublasLt -lnvblas -lcusparse -lcudart -lcudadevrt")
-  find_package(CUDA 10.1)
-  
   if(HIOP_CUDA_LIB_DIR)
-    set(HIOP_CUDA_LIBRARIES "-L${HIOP_CUDA_LIB_DIR} ${HIOP_CUDA_LIBRARIES}")
+    target_link_libraries(hiop_cuda INTERFACE "-L${HIOP_CUDA_LIB_DIR}")
   else()
-    set(HIOP_CUDA_LIBRARIES "-L${CUDA_TOOLKIT_ROOT_DIR}/lib64 ${HIOP_CUDA_LIBRARIES}")
+    target_link_libraries(hiop_cuda INTERFACE "-L${CUDA_TOOLKIT_ROOT_DIR}")
   endif(HIOP_CUDA_LIB_DIR)
 
   # on some systems RHEL Nvidia toolkit puts cuBlas in different locations
   # user can setup/overwrite the lib directory
   if(HIOP_CUBLAS_LIB_DIR)
-    set(HIOP_CUDA_LIBRARIES "-L${HIOP_CUBLAS_LIB_DIR} ${HIOP_CUDA_LIBRARIES}")
+    target_link_libraries(hiop_cuda INTERFACE "-L${HIOP_CUBLAS_LIB_DIR}")
   endif(HIOP_CUBLAS_LIB_DIR)
 
   if(HIOP_CUDA_INCLUDE_DIR)
-    set(HIOP_CUDA_INCLUDE "${HIOP_CUDA_INCLUDE_DIR}")
+    target_include_directories(hiop_cuda INTERFACE "${HIOP_CUDA_INCLUDE_DIR}")
   else()
-    set(HIOP_CUDA_INCLUDE "${CUDA_INCLUDE_DIRS}")
+    target_include_directories(hiop_cuda INTERFACE "${CUDA_INCLUDE_DIR}")
```

```suggestion
    target_include_directories(hiop_cuda INTERFACE ${CUDA_TOOLKIT_INCLUDE})
```
Use variable provided by `find_package(CUDA)`.

---

I think you should also lose `" "` around include dir variables.